dark/diamond/hsp.py

```diff
@@ -1,150 +1,202 @@
+        subject.
+    @param subjectEnd: The 0-based C{int} end offset of the match in the
+        subject.
+    @param queryStart: The 0-based C{int} start offset of the match in the
+        query.
+    @param queryEnd: The 0-based C{int} end offset of the match in the query.
+    @param queryStartInSubject: The 0-based C{int} offset of where the query
+        starts in the subject.
+    @param queryEndInSubject: The 0-based C{int} offset of where the query
+        ends in the subject.
+    @param hsp: The HSP C{dict} passed to normalizeHSP.
+    @param queryLen: the C{int} length of the query sequence.
+    @param subjectGaps: the C{int} number of gaps in the subject.
+    @param queryGaps: the C{int} number of gaps in the query.
+    @param localDict: A C{dict} of local variables from our caller (as
+        produced by locals()).
     """
-    Examine the sense of an HSP and return information about where the
-    read and the alignment (match) begin and end.  Return a dict with keys
-    that allow the read and the alignment to be displayed relative to the
-    hit orientation (i.e., with start < stop for both the read and the
-    match). The returned read indices are offsets into the hit. I.e.,
-    they indicate where on the hit the read lies.
+    # Subject indices must always be ascending.
+    if subjectStart >= subjectEnd:
+        _debugPrint(hsp, queryLen, localDict, 'subjectStart >= subjectEnd')
+
+    subjectMatchLength = subjectEnd - subjectStart
+    queryMatchLength = queryEnd - queryStart
+
+    # Sanity check that the length of the matches in the subject and query
+    # are identical, taking into account gaps in either (indicated by '-'
+    # characters in the match sequences, as returned by DIAMOND).
+    subjectMatchLengthWithGaps = subjectMatchLength + subjectGaps
+    queryMatchLengthWithGaps = queryMatchLength + queryGaps
+    if subjectMatchLengthWithGaps != queryMatchLengthWithGaps:
+        _debugPrint(hsp, queryLen, localDict,
+                    'Including gaps, subject match length (%d) != Query match '
+                    'length (%d)' % (subjectMatchLengthWithGaps,
+                                     queryMatchLengthWithGaps))
+
+    if queryStartInSubject > subjectStart:
+        _debugPrint(hsp, queryLen, localDict,
+                    'queryStartInSubject (%d) > subjectStart (%d)' %
+                    (queryStartInSubject, subjectStart))
+    if queryEndInSubject < subjectEnd:
+        _debugPrint(hsp, queryLen, localDict,
+                    'queryEndInSubject (%d) < subjectEnd (%d)' %
+                    (queryEndInSubject, subjectEnd))
+
+
+def normalizeHSP(hsp, queryLen, diamondTask):
+    """
+    Examine an HSP and return information about where the query and the
+    alignment (match) begin and end.  Return a dict with keys that allow
+    the query and the alignment to be displayed relative to the subject
+    orientation (i.e., with start < stop for both the query and the
```

Comment:
Can we use `match` rather than `alignment`? That would seem to be more consistent with the rest of our language.